### PR TITLE
Added common pattern must to deprecation.

### PR DESCRIPTION
**Pull request by tkrop**
> After the last API guild meeting, I'm unsure, if we should keep the paragraph like this or change it to the MUST/SHOULD headline style. What do you think? (@codekultur  #91)

How is the opinion of the others?

### CODE REVIEW

File: deprecation/Deprecation.md [hunk 1]
```diff
@@ -0,0 +1,60 @@
+#Deprecation
+
+Sometimes it is necessary to phase out an API endpoint (or version). I.e. this
+may be necessary if a field is no longer supported in the result or a whole
+business functionality behind an endpoint has to be shut down. There are
+multiple other reasons as well.
```
Comment (fmueller) [inline deprecation/Deprecation.md]: I'd rephrase: There are many other reasons as well.

File: deprecation/Deprecation.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+endpoint. Producers should help consumers to migrate to a potential new
+endpoint (i.e. by providing a migration manual). After all clients are
+migrated, the producer may shut down the deprecated API.
+
+## {{ book.must }} External Partners Must Agree on Lifecycle Timespans
+
+If the API is consumed by any external partner, the producer must define a
+reasonable timespan that the API will be maintained after the producer has
+announced deprecation. The external partner (client) must agree to this
+minimum after-deprecation-lifespan before he starts using the API.
```
Comment (fmueller) [inline deprecation/Deprecation.md]: What if external client could not agree with producer on lifecycle span? Announce it afterwards and try to reach consensus - then the last sentence of this rule is somehow confusing.

File: deprecation/Deprecation.md [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+## {{ book.must }} Obtain Approval of Clients
+
+Before shutting down an API (or version of an API) the producer must make
+sure, that all clients have given their consent to shut down the
+endpoint. Producers should help consumers to migrate to a potential new
+endpoint (i.e. by providing a migration manual). After all clients are
+migrated, the producer may shut down the deprecated API.
+
+## {{ book.must }} External Partners Must Agree on Lifecycle Timespans
```
Comment (fmueller) [inline deprecation/Deprecation.md]: ...agree on deprecation timespan

File: deprecation/Deprecation.md [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+If the API is consumed by any external partner, the producer must define a
+reasonable timespan that the API will be maintained after the producer has
+announced deprecation. The external partner (client) must agree to this
+minimum after-deprecation-lifespan before he starts using the API. 
+
+## {{ book.must }} Reflect Deprecation in API Definition
+
+API deprecation must reflect in the producers OpenAPI definition. If a method
```
Comment (fmueller) [inline deprecation/Deprecation.md]: ... must be part of OpenAPI definition

File: deprecation/Deprecation.md [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+on a path, a whole path or even a whole API endpoint (multiple paths) should
+be deprecated, the producers must set `deprecated=true` on each method / path
+element that will be deprecated (OpenAPI 2.0 only allows you to define
+deprecation on this level). If deprecation should happen on a finer level
```
Comment (fmueller) [inline deprecation/Deprecation.md]: ... should happen on a more fine-grained level

File: SUMMARY.md [hunk 6]
```diff
@@ -16,6 +16,7 @@
 * [Common Data Objects](common-data-objects/CommonDataObjects.md)
 * [Common Headers](headers/CommonHeaders.md)
 * [Proprietary Headers](headers/ProprietaryHeaders.md)
+* [Deprecation](deprecation/Deprecation.md)
```
Comment (fmueller) [inline SUMMARY.md]: I'd put it under Compatibility as it describes what happens when compatibility can not be guaranteed anymore.

File: deprecation/Deprecation.md [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+(i.e. query parameter, payload etc.), the producer should set
+`deprecated=true` on the affected method / path element and add further
+explanation to the `description` section. 
+
+If `deprecated` is set to `true`, the producer must describe what clients
+should use instead and when the API will be shut down in the `description`
```
Comment (fmueller) [inline deprecation/Deprecation.md]: What if the deprecated business functionality is not provided anymore in any sense?

File: deprecation/Deprecation.md [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+section of the API definition. 
+
+## {{ book.should }} Monitor Usage of Deprecated APIs
+
+Producers should monitor usage of deprecated APIs.
```
Comment (fmueller) [inline deprecation/Deprecation.md]: ... and when they fade out.

File: deprecation/Deprecation.md (hunk 1 above)
Comment (codekultur) [inline deprecation/Deprecation.md]: I'll change it..

File: deprecation/Deprecation.md [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+field. When adding the `Warning` header, the `warn-code` must be `299` and the
+`warn-text` should be in form of *"The path/operation/parameter/... {name} is
+deprecated and will be removed by {date}. Please see {link} for details."*
+with a link to a documentation describing why the API is no longer supported
+in the current form and what clients should do about it. Adding the `Warning`
+header is not sufficient to gain client consent to shut down an API.
+
+## {{ book.should }} Add Monitoring for Warning Header
+
+Clients should monitor the `Warning` header in HTTP responses.
```
Comment (fmueller) [inline deprecation/Deprecation.md]: ... to see if an API will be deprecated in future.

File: deprecation/Deprecation.md (hunk 2 above)
Comment (codekultur) [inline deprecation/Deprecation.md]: The trick is that the external partner has to agree with the producer before he is allowed to use the API (i.e. by terms & conditions), just like you do when you use the Salesforce, eBay, AWS API...

File: deprecation/Deprecation.md (hunk 3 above)
Comment (codekultur) [inline deprecation/Deprecation.md]: sounds good. I'll change it

File: deprecation/Deprecation.md (hunk 5 above)
Comment (codekultur) [inline deprecation/Deprecation.md]: I'll change it

File: deprecation/Deprecation.md (hunk 2 above)
Comment (fmueller) [inline deprecation/Deprecation.md]: Right but when we add this chapter to the guidelines there are already APIs for external partners available without negotiate this in Terms&Conditions. For new APIs this makes totally sense.

File: SUMMARY.md (hunk 6 above)
Comment (codekultur) [inline SUMMARY.md]: We discussed this option as well.. We put it in the end, as it reflects the API lifespan, but I agree it would also fit under Compatibility...

Comment (fmueller) [inline SUMMARY.md]: Ok, then we leave it there.

File: deprecation/Deprecation.md (hunk 7 above)
Comment (codekultur) [inline deprecation/Deprecation.md]: I could add "if there is an alternative", but this would make the sentence more complex and I think it's clear to everyone, isn't it?

File: deprecation/Deprecation.md (hunk 9 above)
Comment (codekultur) [inline deprecation/Deprecation.md]: I'll add this

File: deprecation/Deprecation.md (hunk 4 above)
Comment (codekultur) [inline deprecation/Deprecation.md]: I'll change it

File: deprecation/Deprecation.md (hunk 8 above)
Comment (codekultur) [inline deprecation/Deprecation.md]: Sorry... Don't understand what you mean... 

"Producers should monitor usage of deprecated APIs and when they fade out." ?!?

File: deprecation/Deprecation.md (hunk 2 above)
Comment (codekultur) [inline deprecation/Deprecation.md]: Hmm... Maybe this is a general discussion: Our guidelines evolve over time. Should we really always consider how the guidelines once looked like and provide migration scenarios, or should we keep the guidelines consistent at one point in time?
I would vote for the second option, as it keeps the guidelines short and easy to understand. But of course it won't give guidance for what to do with existing APIs. I would say: Try to get them compliant to the guidelines, as soon as possible. In this case, there weren't any rules for deprecation before, so the external clients have no guarantee at all so far...

File: SUMMARY.md (hunk 6 above)
Comment (tkrop) [inline SUMMARY.md]: I originally voted for compatibility, too. But my main idea for setting it below, was making it less prominent.

File: deprecation/Deprecation.md (hunk 2 above)
Comment (tkrop) [inline deprecation/Deprecation.md]: I think, than it is a legal problem too. If external consumers have not agreed upfront, we may be in trouble, because we may be forced to produce infinitely on this contract, except there is a possibility for canceling it.

File: deprecation/Deprecation.md (hunk 8 above)
Comment (tkrop) [inline deprecation/Deprecation.md]: May be we should add 'until it can be shut down.'

Comment (fmueller) [inline deprecation/Deprecation.md]: I mean monitoring the timespan til the endpoint/api is shut down.

File: deprecation/Deprecation.md (hunk 7 above)
Comment (fmueller) [inline deprecation/Deprecation.md]: Ok

File: SUMMARY.md (hunk 6 above)
Comment (fmueller) [inline SUMMARY.md]: Right, making it less prominent is also a good idea as this should be an edge case.

File: deprecation/Deprecation.md (hunk 2 above)
Comment (fmueller) [inline deprecation/Deprecation.md]: Both of you are right. Let's leave it as it is.

File: deprecation/Deprecation.md (hunk 8 above)
Comment (codekultur) [inline deprecation/Deprecation.md]: OK. Changed the sentence...

**Comment by fmueller** [comment]
👍

**Comment by tkrop** [comment]
:+1:
